### PR TITLE
[feature](function) support ip functions isipv4string and isipv6string

### PR DESCRIPTION
**Pull request by sjyango**
## Proposed changes

Issue Number: #20584

doris output:
image
image

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run feut

**Comment by nanfeng1999** [comment]
run buildall

**Comment by nanfeng1999** [comment]
run arm

**Comment by nanfeng1999** [comment]
run arm

**Comment by nanfeng1999** [comment]
run arm

**Comment by nanfeng1999** [comment]
run buildall

**Comment by nanfeng1999** [comment]
run buildall

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run pipelinex_p0

File: be/src/vec/runtime/ipv6_value.h [hunk 1]
```diff
@@ -65,6 +65,11 @@ class IPv6Value {
         return {buf, len};
     }
 
+    static bool is_valid_string(const std::string& ipv6_str) {
+        vectorized::IPv6 value;
+        return from_string(value, ipv6_str);
```
Comment (amorynan) [inline be/src/vec/runtime/ipv6_value.h:70]: does from_string costs more time than parseIpV4?

File: be/src/vec/functions/function_ip.h [hunk 2]
```diff
@@ -348,4 +348,108 @@ class FunctionIPv6NumToString : public IFunction {
...
+class FunctionIsIPv6String : public IFunction {
+private:
+    Status execute_type(Block& block, const ColumnWithTypeAndName& argument, size_t result) const {
+        const ColumnPtr& column = argument.column;
+
+        if (const auto* col_src = typeid_cast<const ColumnString*>(column.get())) {
+            auto col_res = ColumnInt8::create();
+
+            for (size_t i = 0; i < col_src->size(); ++i) {
+                auto ipv6_str = col_src->get_data_at(i).to_string();
```
Comment (amorynan) [inline be/src/vec/functions/function_ip.h:412]: can u just get data char pointer to check? no need call to_string() which is costs much

File: be/src/vec/functions/function_ip.h [hunk 3]
```diff
@@ -348,4 +348,108 @@ class FunctionIPv6NumToString : public IFunction {
...
+                if (ipv6_str.size() > IPV6_MAX_TEXT_LENGTH ||
+                    !IPv6Value::is_valid_string(ipv6_str)) {
+                    col_res->insert_value(0);
+                } else {
+                    col_res->insert_value(1);
+                }
+            }
+
+            DCHECK_EQ(col_res->size(), col_src->size());
```
Comment (amorynan) [inline be/src/vec/functions/function_ip.h:421]: this DCHECK should move to begin, not here when not do anything after some unpredict thing happend.

File: be/src/vec/functions/function_ip.h [hunk 4]
```diff
@@ -348,4 +348,108 @@ class FunctionIPv6NumToString : public IFunction {
...
+    }
+};
+
+class FunctionIsIPv6String : public IFunction {
+private:
+    Status execute_type(Block& block, const ColumnWithTypeAndName& argument, size_t result) const {
+        const ColumnPtr& column = argument.column;
+
+        if (const auto* col_src = typeid_cast<const ColumnString*>(column.get())) {
+            auto col_res = ColumnInt8::create();
```
Comment (amorynan) [inline be/src/vec/functions/function_ip.h:409]: I think here is mistake for boolean value which now we make ColumnUInt8 present boolean!

File: be/src/vec/functions/function_ip.h [hunk 5]
```diff
@@ -348,4 +348,108 @@ class FunctionIPv6NumToString : public IFunction {
...
+        return execute_type(block, argument, result);
+    }
+};
+
+class FunctionIsIPv6String : public IFunction {
+private:
+    Status execute_type(Block& block, const ColumnWithTypeAndName& argument, size_t result) const {
+        const ColumnPtr& column = argument.column;
+
+        if (const auto* col_src = typeid_cast<const ColumnString*>(column.get())) {
```
Comment (amorynan) [inline be/src/vec/functions/function_ip.h:408]: here if columns is  ColumnNullable<ColumnString> ?

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run p0

File: be/src/vec/functions/function_ip.h [hunk 6]
```diff
@@ -348,4 +348,138 @@ class FunctionIPv6NumToString : public IFunction {
...
+            size_t col_size = nullable_src->size();
+            auto col_res = ColumnUInt8::create(col_size, 0);
+            auto null_map = ColumnUInt8::create(col_size, 0);
+
+            for (size_t i = 0; i < col_size; ++i) {
+                if (nullable_src->is_null_at(i)) {
+                    null_map->get_data()[i] = 1;
+                } else {
+                    StringRef ipv4_str = nullable_src->get_data_at(i);
+                    if (IPv4Value::is_valid_string(ipv4_str.data, ipv4_str.size)) {
```
Comment (amorynan) [inline be/src/vec/functions/function_ip.h:366]: why not deal with invalid string ?

File: be/src/vec/functions/function_ip.h [hunk 7]
```diff
@@ -348,4 +348,138 @@ class FunctionIPv6NumToString : public IFunction {
...
+        const ColumnPtr& column = argument.column;
+
+        if (const auto* nullable_src = typeid_cast<const ColumnNullable*>(column.get())) {
+            size_t col_size = nullable_src->size();
+            auto col_res = ColumnUInt8::create(col_size, 0);
+            auto null_map = ColumnUInt8::create(col_size, 0);
+
+            for (size_t i = 0; i < col_size; ++i) {
+                if (nullable_src->is_null_at(i)) {
+                    null_map->get_data()[i] = 1;
```
Comment (amorynan) [inline be/src/vec/functions/function_ip.h:363]: manipulate with data directly, same with col_res

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run p0

**Comment by sjyango** [comment]
run p0

**Comment by sjyango** [comment]
run buildall

**Comment by sjyango** [comment]
run feut

**Review by amorynan — APPROVED** [review]
LGTM

**Review by yiguolei — APPROVED** [review]

**Pull request merged**